File: LC/Medium/567.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        count = Counter(s1)
        window_start = 0
        
        curr = defaultdict(int)
        for window_end in range(len(s2)):
            curr[s2[window_end]] += 1
            if curr == count:
                return True
            
            while window_end - window_start + 2 > len(s1):
                curr[s2[window_start]] -= 1
                if curr[s2[window_start]] == 0:
                    del curr[s2[window_start]]
                window_start += 1
        
        return False

File: LC/C++/Medium/567.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if (s1.size() > s2.size()) return false;

        std::array<int, 26> target{};
        std::array<int, 26> window{};

        for (char c : s1)
            target[c - 'a']++;

        int l = 0;
        for (int r = 0; r < s2.size(); ++r) {
            window[s2[r] - 'a']++;

            if (r >= s1.size()) {
                window[s2[l] - 'a']--;
                l++;
            }

            if (window == target)
                return true;
        }

        return false;
    }
};